### PR TITLE
Implement anonymous worksheets and bundles

### LINKED ISSUE #400: Anonymous worksheets and bundles
When submitting to a double-blind conference, it would be nice to be able to include a link to a CodaLab worksheet without revealing the user identity.  Worksheets and bundles should have an anonymous flag such that when set, people can still read the bundle, but the user is not revealed.

### LINKED ISSUE #769: REST API commands fetch more data than necessary
This makes for noticeable slowdowns on some operations and data. We should spend some time designing query parameters for specifying what fields the client needs. 

For example:

> Often when we resolve bundle specs through GET /bundles?specs=a,b,c we don't need anything more than the UUIDs.

> The responses of GET /bundles/<uuid> currently contains more information than necessary, in particular, the full bundle_infos of the children of the bundle are included in the responses, when really only their names are needed.

### PR DESCRIPTION
**Pull request by kashizui**
- [x] add anonymous flag and make sure it goes all the way to the user
- [x] finalize desired behavior of anonymous flag (#400) 
- [x] hide owner/groups of an anonymous bundle to users with permission < 2.
- [x] allow owner (and users with all permission?) to make a bundle anonymous through REST API
- [x] metadata editor in CLI should allow owners to change anonymous flag
- [x] ~support recursive anonymizing / anonymizing entire worksheet and its attached bundles in CLI~ not in this PR
- [x] implement permissions policy
- [x] new bundles inherit anonymous flag from parent worksheet

Also adds support for "include" query parameter on bundle and worksheet APIs to allow clients to choose which related resources to include in the response. This should make CLI much snappier.

To be merged with codalab/codalab-worksheets#332
Fixes #400
Fixes #769

### CODE REVIEW

**Comment by kashizui** [comment]
Ready for review! @percyliang @fabeschan @maxwang7

**Comment by kashizui** [comment]
A reminder to review this please :)

**Comment by fabeschan** [comment]
Are the terms `private` and `anonymous` interchangeable?

File: codalab/client/json_api_client.py [hunk 1]
```diff
@@ -162,12 +162,22 @@ def _get_resource_path(resource_type, resource_id=None, relationship=None):
...
+
+        :param dict[str, *] params: map from query parameter key to value
+        :rtype: list[(str, str)]
         """
         if params is None:
             return None
 
         result = []
+
+        # `include` query parameter is a comma-spearated list, as defined by:
```
Comment (fabeschan) [inline codalab/client/json_api_client.py]: typo comma-separated

File: codalab/lib/bundle_cli.py [hunk 2]
```diff
@@ -1357,6 +1354,10 @@ def do_push_image_command(self, args):
             Commands.Argument('-n', '--name', help='Change the bundle name (format: %s).' % spec_util.NAME_REGEX.pattern),
             Commands.Argument('-T', '--tags', help='Change tags (must appear after worksheet_spec).', nargs='*'),
             Commands.Argument('-d', '--description', help='New bundle description.'),
+            Commands.Argument('--anonymous', help='Set bundle to be anonymous (identity of the owner will NOT \n'
+                              'be visisble to users without \'all\' permission on the bundle).',
```
Comment (fabeschan) [inline codalab/lib/bundle_cli.py]: typo visible

File: codalab/lib/bundle_cli.py [hunk 3]
```diff
@@ -2202,13 +2210,19 @@ def change_current_worksheet(self, client, worksheet_uuid, verbose=False):
             Commands.Argument('-T', '--tags', help='Change tags (must appear after worksheet_spec).', nargs='*'),
             Commands.Argument('-o', '--owner-spec', help='Change owner of worksheet.'),
             Commands.Argument('--freeze', help='Freeze worksheet to prevent future modification (PERMANENT!).', action='store_true'),
+            Commands.Argument('--anonymous', help='Set worksheet to be anonymous (identity of the owner will NOT \n'
+                                                  'be visisble to users without \'all\' permission on the worksheet).',
```
Comment (fabeschan) [inline codalab/lib/bundle_cli.py]: typo ditto

**Comment by kashizui** [comment]
yup private == anonymous -- sorry I should probably make that consistent in the code comments.

**Review by fabeschan — APPROVED** [review]
Looks good to me! I think we really should try to get this in before everybody goes off for the summer.

**Comment by kashizui** [comment]
Ah sorry actually `PrivateBundle` refers to a bundle that a user has `none` access to. (all fields are hidden).

**Pull request merged**